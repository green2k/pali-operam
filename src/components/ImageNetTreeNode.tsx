import * as React from 'react'
import * as R from 'ramda'

interface IImageNetTreeNode {
  id: number
  name: string
  size: number
  children: IImageNetTreeNode[]
}

interface IIImageNetTreeNodeProps {
  tree: IImageNetTreeNode
}

// Nice & beauty ASCI visualization of a tree-node
const buildNodePrefix: (level: number) => string = level => {
  return new Array(level * 5 + 1).join(' ') + ' └──'
}

// Build a node-suffix by it's total count of leafs
const buildNodeSuffix: (countLeafs: number) => string = countLeafs => {
  return countLeafs > 0 ? '(' + String(countLeafs) + ')' : ''
}

// Returns the number of leafs contained in `tree`
export const containsLeaves: (tree: IImageNetTreeNode) => number = (tree) => {
  return R.reduce((acc: number, el: IImageNetTreeNode): number => (
    (el.size === 0) ? acc + 1 : acc + containsLeaves(el)
  ), 0 as number, tree.children)
}

export default class ImageNetTreeNode extends React.PureComponent<IIImageNetTreeNodeProps> {

  // Renders a single tree-node
  private renderNode(tree: IImageNetTreeNode, level: number): JSX.Element | null {
    const countLeafs = containsLeaves(tree)

    if (countLeafs === 0 && tree.size > 0) { // Intermediate node with no leafs
      return null
    }

    return (
      <div key={tree.id}>
        <pre>{buildNodePrefix(level)} {tree.name} {buildNodeSuffix(countLeafs)}</pre>
        <div>
          { tree.children.map(child => this.renderNode(child, level + 1)) }
        </div>
      </div>
    )
  }

  render() {
    return this.renderNode(this.props.tree, 0)
  }
}

/*
 * TODO: Extract some functions into /helpers/ImageNetTree.js
 */
