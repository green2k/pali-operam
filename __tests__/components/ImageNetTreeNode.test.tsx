import * as React from 'react'
import * as Renderer from 'react-test-renderer'

import ImageNetTreeNode, { containsLeaves } from '../../src/components/ImageNetTreeNode'

const TREE_1 = {
  id: 1,
  name: 'Top parent',
  size: 5,
  children: [
    {id: 2, name: 'Child1', size: 0, children: []},
    {id: 3, name: 'Intermediate1', size: 3, children: [
      {id: 4, name: 'Child1Child1', size: 0, children: []},
      {id: 5, name: 'Child1Child2', size: 0, children: []},
      {id: 6, name: 'Child1Child3', size: 0, children: []},
    ]},
    {id: 7, name: 'Child3', size: 0, children: []},
  ]
}

describe('<ImageNetTreeNode>', () => {
  it('renders', () => {
    const actual = Renderer
      .create(<ImageNetTreeNode tree={TREE_1} />)
      .toJSON()
    expect(actual).toMatchSnapshot()
  })
})

describe('containsLeaves(..)', () => {
  it('Returns a correct number of leaves', () => {
    const actual = containsLeaves(TREE_1)
    expect(actual).toBe(5)
  })
})
