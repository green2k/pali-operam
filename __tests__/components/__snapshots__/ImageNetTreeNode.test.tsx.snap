// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageNetTreeNode> renders 1`] = `
<div>
  <pre>
     └──
     
    Top parent
     
    (5)
  </pre>
  <div>
    <div>
      <pre>
              └──
         
        Child1
         
        
      </pre>
      <div />
    </div>
    <div>
      <pre>
              └──
         
        Intermediate1
         
        (3)
      </pre>
      <div>
        <div>
          <pre>
                       └──
             
            Child1Child1
             
            
          </pre>
          <div />
        </div>
        <div>
          <pre>
                       └──
             
            Child1Child2
             
            
          </pre>
          <div />
        </div>
        <div>
          <pre>
                       └──
             
            Child1Child3
             
            
          </pre>
          <div />
        </div>
      </div>
    </div>
    <div>
      <pre>
              └──
         
        Child3
         
        
      </pre>
      <div />
    </div>
  </div>
</div>
`;
